rsc/client/valve/replay.go

```diff
@@ -0,0 +1,95 @@
+package valve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var (
+	ErrDownloadRelpay = fmt.Errorf("failed to download the dota2 replay")
+)
+
+type replay struct {
+	httpClient *http.Client
+}
+
+func NewReplay(httpClient *http.Client) *replay {
+	return &replay{
+		httpClient: httpClient,
+	}
+}
+
+// Download
+// downloading the replay file is downloaded directly to the
+// Valve replay proxy using the replayURL.
+//
+// replayURL contains the full raw URI to download the replay to the Valve replay proxy,
+// for example http://replay153.valve.net/570/7132230434_1635105612.dem.bz2
+// Based on that value, we can download the replay DotA2 file directly using that URL.
+//
+// destination parameter specifies the file destination to save the downloaded replay file.
+func (r *replay) Download(ctx context.Context, replayURL string, destination string) error {
```

Comment:
this is just a note: I am half expecting we would refactor this function a bit, with regards to function parameters and returns, but I might be wrong so we can keep this for now.